libs/framework/src/bundle_context.c

```diff
@@ -1181,78 +1182,70 @@ static void celix_bundleContext_useServiceWithOptions_2_UseServiceTracker(void *
     assert(celix_framework_isCurrentThreadTheEventLoop(d->ctx->framework));
 
-    d->called = celix_serviceTracker_useHighestRankingService(d->svcTracker, d->opts->filter.serviceName, d->opts->callbackHandle, d->opts->use, d->opts->useWithProperties, d->opts->useWithOwner);
-}
-
-static void celix_bundleContext_useServiceWithOptions_3_CloseServiceTracker(void *data) {
-    celix_bundle_context_use_service_data_t* d = data;
-    assert(celix_framework_isCurrentThreadTheEventLoop(d->ctx->framework));
-
-    celix_service_tracker_t *tracker = d->svcTracker;
-    d->svcTracker = NULL;
-
-    celix_serviceTracker_destroy(tracker);
+    d->called = celix_serviceTracker_useHighestRankingService(d->svcTracker, d->opts->filter.serviceName, 0, d->opts->callbackHandle, d->opts->use, d->opts->useWithProperties, d->opts->useWithOwner);
 }
 
 bool celix_bundleContext_useServiceWithOptions(
         celix_bundle_context_t *ctx,
         const celix_service_use_options_t *opts) {
-    if (opts == NULL) {
+    if (opts == NULL || opts->filter.serviceName == NULL) {
         return false;
     }
 
     celix_bundle_context_use_service_data_t data = {0};
     data.ctx = ctx;
     data.opts = opts;
+    bool called = false;
 
     if (celix_framework_isCurrentThreadTheEventLoop(ctx->framework)) {
         // Ignore timeout: blocking the event loop prevents any progress to be made
         celix_bundleContext_useServiceWithOptions_1_CreateServiceTracker(&data);
         celix_bundleContext_useServiceWithOptions_2_UseServiceTracker(&data);
-        celix_bundleContext_useServiceWithOptions_3_CloseServiceTracker(&data);
+        celix_serviceTracker_destroy(data.svcTracker);
         return data.called;
     }
 
     long eventId = celix_framework_fireGenericEvent(ctx->framework, -1, celix_bundle_getId(ctx->bundle), "create service tracker for celix_bundleContext_useServiceWithOptions", &data, celix_bundleContext_useServiceWithOptions_1_CreateServiceTracker, NULL, NULL);
     celix_framework_waitForGenericEvent(ctx->framework, eventId);
 
-    celix_framework_waitForEmptyEventQueue(ctx->framework); //ensure that a useService wait if a listener hooks concept, which triggers an async service registration
 
-    struct timespec startTime = celix_gettime(CLOCK_MONOTONIC);
-    bool useServiceIsDone = false;
-    bool called = false;
-    do {
-        eventId = celix_framework_fireGenericEvent(ctx->framework, -1, celix_bundle_getId(ctx->bundle), "use service tracker for celix_bundleContext_useServiceWithOptions", &data, celix_bundleContext_useServiceWithOptions_2_UseServiceTracker, NULL, NULL);
-        celix_framework_waitForGenericEvent(ctx->framework, eventId);
+    if(opts->flags & CELIX_SERVICE_USE_DIRECT) {
+        celix_framework_waitUntilNoPendingRegistration(ctx->framework); // TL;DR make "service on demand" pattern work. Try comment it out and run CelixBundleContextServicesTests.UseServiceOnDemandDirectlyWithAsyncRegisterTest
```

Comment:
IMO this can be removed. As long as we document that `CELIX_SERVICE_USE_DIRECT` will not work with the "service on demand" pattern. 

This should also make it possible to work directly on service registry to prevent even more overhead (future, if we want to and this will be a significant refactoring).

Reply:
@PengZheng WDYT?

Reply:
Now `CELIX_SERVICE_USE_DIRECT` will not support SOD by default, unless explicitly requested by user.

Since last mails, I've thought about this. I see one potential problem with direct registry usage: global lock contention, i.e. users interested in different services will affect each other. I have the following scenario in mind:

1. Create one tracker (on bundle start maybe).
2. Use service via the same tracker concurrently from multiple threads.
3. Destroy the tracker (on bundle stop maybe).

By design, users of different services use different trackers, global lock contention is reduced into per tracker lock contention. Furthermore, performance of concurrent use of the same tracker can be further improved by using read-write lock instead of the tracker lock. 

However, current tracker implementation is sub-optimal: currently there is no way to obtain opaque tracker pointer, only the tracker id is returned. The translation between tracker id and opaque tracker pointer involves bundle context lock. Thus our design only reduces global lock contention into per bundle lock contention, not the optimal per tracker lock contention. If we can get opaque tracker pointer from tracker id (service tracker is still created on the event loop), then `celix_serviceTracker_useHighestRankingService`/`celix_serviceTracker_useHighestRankingService` can help us achieving optimal performance. Since no event loop is involved in `celix_serviceTracker_useHighestRankingService`, its recursive usage (using a service itself calling useService) together with async API may not hurt us any more.  IIRC, the intention of original OSGi service tracker is to avoid external synchronization implemented by user.  By re-exposing `celix_serviceTracker_useHighestRankingService` as public API, we can recovery its original design purpose.

I planed to discuss it with you after the release happens, since it involves public API change bigger than this PR. 

PS: Currently I'm working on conan support. Hopefully, a PR will arrive this week. @pnoltes

Reply:
That is an interesting scenario and would IMO  indeed make it possible to make better performance use service calls without breaking api or putting extra stress on the service registry.

I also agree with making parts of the service tracker more accessible.  

The idea of returning "service tracker ids" instead of pointers to a service tracker has to do with ownership and difficulties in C to arrange or share ownership. IMO is was too easy to forward service tracker (or service registration) pointers without providing control of the service tracker lifecycle or checks if the service tracker is still valid. 
But indeed the downside is that the service tracker api is now hidden and we could provide the api  through the bundle context adds additional overhead. 

Another note: The service tracker was designed for multi threading and has not yet been updated for the fact that is should only be created, destroyed and get services update from a single Celix event thread. I have not look into this yet, but this could also mean that service trackers should be able to operate without mutexes, but with atomics or even lock-free constructs instead.

Reply:
Yes, I can see the rational behind ids. That's why I made `celix_serviceRegistry_unregisterService` MT-safe even when concurrently called with the same serviceId from multiple threads. The second point is almost true except that `celix_serviceRegistry_unregisterService` can also be called by `celix_bundleContext_cleanup`, which is in turn called by `celix_framework_stopBundleOnANonCelixEventThread`.  By the comments in `celix_bundleContext_cleanup`, it may be viewed as an assertion failure? Indeed, for this almost-one-writer-multiple-reader case, further optimization is possible.

For the mentioned scenario, an additional `celix_bundleContext_getServiceTrackerFromId` might be enough. However, the semantics of `celix_serviceTracker_useServices` is not consistent with that of `celix_service_tracking_options_t`, especially the callbacks part.

PS: let me have a try of my commit right ;p